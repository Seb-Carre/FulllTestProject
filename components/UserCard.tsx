import {
  Dimensions,
  Image,
  Linking,
  Pressable,
  StyleSheet,
  View,
} from 'react-native';

import Button from './Button';
import Checkbox from './Checkbox';
import CustomText from './CustomText';
import React from 'react';
import {User} from '../types';

type UserCardProps = {
  user: User;
  setSelected: (user: User) => void;
  selected: boolean;
};

/**
 * Functional Component to show the user card
 * @param {User} user
 * @param {(user: User) void} setSelected
 * Put the user to the selected list from the parent component
 * @param {boolean} selected
 * Manage the status of the checkbox inside the card
 */
export default function UserCard({user, setSelected, selected}: UserCardProps) {
  const onPressViewProfile = () => {
    // if the user has a html url
    if (user?.html_url) {
      // we are going to open this url in a web browser
      Linking.openURL(user.html_url);
    }
  };
  const onPressSelectedUser = () => {
    // using the setSelected method from parent component to manage the state
    setSelected(user);
  };
  return (
    <View style={styles.container}>
      <View style={styles.infosUserContainer}>
        <Pressable
          onPress={onPressSelectedUser}
          style={styles.checkboxPosition}>
          <Checkbox mode={selected ? 'active' : 'none'} />
        </Pressable>
        <View style={styles.infoUser}>
          <Image style={styles.tinyLogo} source={{uri: user?.avatar_url}} />
          <CustomText textType="title">{user?.id}</CustomText>
          <CustomText textType="title">{user?.login}</CustomText>
        </View>
      </View>
      <View style={styles.buttonContainer}>
        <Button onPress={onPressViewProfile}>View profile</Button>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: Dimensions.get('screen').width,
    borderRadius: 20,
    marginBottom: 16,
    backgroundColor: 'lightgray',
  },
  infosUserContainer: {flex: 1, padding: 20, flexDirection: 'row'},
  buttonContainer: {margin: 15, alignItems: 'center'},
  checkboxPosition: {
    position: 'absolute',
    padding: 25,
    zIndex: 1,
  },
  infoUser: {flex: 1, alignItems: 'center'},
  tinyLogo: {
    width: 70,
    height: 70,
    borderRadius: 35,
  },
});
