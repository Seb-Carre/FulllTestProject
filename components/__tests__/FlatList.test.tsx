import {FlatList} from 'react-native';
import React from 'react';
import {User} from '../../types';
import UserCard from '../UserCard';
import {render} from '@testing-library/react-native';

const users: User[] = [
  {
    login: 'sebastien',
    id: 404,
    avatar_url: 'https://reactnative.dev/img/tiny_logo.png',
    html_url: 'https://github.com/Seb-Carre',
  },
  {
    login: 'eric',
    id: 500,
    avatar_url: 'https://reactnative.dev/img/tiny_logo.png',
    html_url: 'https://github.com/eric',
  },
];

it('render the Flatlist with our userCard as snapShot excepted', () => {
  const {toJSON} = render(
    <FlatList
      data={users} // insert the fake data we made
      keyExtractor={(user, index) => index.toString()}
      renderItem={({item}) => (
        <UserCard
          testID="userCard"
          user={item}
          setSelected={() => {}}
          selected={true}
        />
      )}
    />,
  );

  expect(toJSON()).toMatchSnapshot();
});

it('render the Flatlist with our userCard and check if our list is render correctly with users', () => {
  const {getByTestId, getByText} = render(
    <FlatList
      data={users}
      testID={'FlatList'}
      keyExtractor={(user, index) => index.toString()}
      renderItem={({item}) => (
        <UserCard user={item} setSelected={() => {}} selected={true} />
      )}
    />,
  );
  const list = getByTestId('FlatList');
  // Check to have our ListFlat, it need to be present of course
  expect(list).not.toBeNull();
  // We have two users, we are checking if the login and id is begin showed for both
  users.forEach(user => {
    expect(getByText(user.login)).not.toBeNull();
    expect(getByText(user.id.toString())).not.toBeNull();
  });
});
