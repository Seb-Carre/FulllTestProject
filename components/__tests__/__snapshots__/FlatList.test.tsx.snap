// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the Flatlist with our userCard as snapShot excepted 1`] = `
<RCTScrollView
  data={
    [
      {
        "avatar_url": "https://reactnative.dev/img/tiny_logo.png",
        "html_url": "https://github.com/Seb-Carre",
        "id": 404,
        "login": "sebastien",
      },
      {
        "avatar_url": "https://reactnative.dev/img/tiny_logo.png",
        "html_url": "https://github.com/eric",
        "id": 500,
        "login": "eric",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "backgroundColor": "lightgray",
            "borderRadius": 20,
            "height": 750,
            "marginBottom": 16,
            "width": "100%",
          }
        }
        testID="App.UserCard"
      >
        <View
          style={
            {
              "flex": 1,
              "flexDirection": "row",
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "padding": 25,
                "position": "absolute",
                "zIndex": 1,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "backgroundColor": "#44a1f8",
                  "borderRadius": 4,
                  "borderWidth": 4,
                  "height": 20,
                  "justifyContent": "center",
                  "width": 20,
                }
              }
              testID="UserCard.CheckBox"
            >
              <Image
                source={
                  {
                    "testUri": "../../../assets/images/check-mark.png",
                  }
                }
                style={
                  {
                    "height": 13,
                    "width": 13,
                  }
                }
                testID="Checkbox.icon"
              />
            </View>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
              }
            }
          >
            <Image
              source={
                {
                  "uri": "https://reactnative.dev/img/tiny_logo.png",
                }
              }
              style={
                {
                  "borderRadius": 35,
                  "height": 70,
                  "width": 70,
                }
              }
            />
            <Text
              style={
                [
                  {
                    "fontSize": 20,
                    "fontWeight": "600",
                  },
                  undefined,
                ]
              }
            >
              404
            </Text>
            <Text
              style={
                [
                  {
                    "fontSize": 20,
                    "fontWeight": "600",
                  },
                  undefined,
                ]
              }
            >
              sebastien
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "margin": 15,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#00a2ff",
                "borderRadius": 10,
                "paddingHorizontal": 50,
                "paddingVertical": 20,
              }
            }
            testID="UserCard.ViewProfileButton"
          >
            <Text
              style={
                [
                  {},
                  {
                    "fontSize": 20,
                  },
                ]
              }
            >
              View profile
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "backgroundColor": "lightgray",
            "borderRadius": 20,
            "height": 750,
            "marginBottom": 16,
            "width": "100%",
          }
        }
        testID="App.UserCard"
      >
        <View
          style={
            {
              "flex": 1,
              "flexDirection": "row",
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "padding": 25,
                "position": "absolute",
                "zIndex": 1,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "backgroundColor": "#44a1f8",
                  "borderRadius": 4,
                  "borderWidth": 4,
                  "height": 20,
                  "justifyContent": "center",
                  "width": 20,
                }
              }
              testID="UserCard.CheckBox"
            >
              <Image
                source={
                  {
                    "testUri": "../../../assets/images/check-mark.png",
                  }
                }
                style={
                  {
                    "height": 13,
                    "width": 13,
                  }
                }
                testID="Checkbox.icon"
              />
            </View>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
              }
            }
          >
            <Image
              source={
                {
                  "uri": "https://reactnative.dev/img/tiny_logo.png",
                }
              }
              style={
                {
                  "borderRadius": 35,
                  "height": 70,
                  "width": 70,
                }
              }
            />
            <Text
              style={
                [
                  {
                    "fontSize": 20,
                    "fontWeight": "600",
                  },
                  undefined,
                ]
              }
            >
              500
            </Text>
            <Text
              style={
                [
                  {
                    "fontSize": 20,
                    "fontWeight": "600",
                  },
                  undefined,
                ]
              }
            >
              eric
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "margin": 15,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#00a2ff",
                "borderRadius": 10,
                "paddingHorizontal": 50,
                "paddingVertical": 20,
              }
            }
            testID="UserCard.ViewProfileButton"
          >
            <Text
              style={
                [
                  {},
                  {
                    "fontSize": 20,
                  },
                ]
              }
            >
              View profile
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
